services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeServiceImpl.kt

```diff
@@ -271,38 +245,97 @@ class HankeServiceImpl(private val hankeRepository: HankeRepository,
+
+        // If there is anything left in the existingYTs map, they have been removed in the incoming data,
+        // so remove them from the entity's list (and thus from the database).
+        // TODO: it should not be a list, but a "bag".  Their order does not matter, and removing things from the list
+        // gets inefficient. Since there are so few entries, this crude solution works, for now.
+        entity.listOfHankeYhteystieto.removeAll(existingYTs.values)
     }
 
-    internal fun addHankeYhteystietoEntitysToList(listOfHankeYhteystiedot: List<HankeYhteystieto>, hankeEntity: HankeEntity, contactType: ContactType) {
-
-        listOfHankeYhteystiedot.forEach { hankeYht ->
-            if (hankeYht.sukunimi.isNotBlank() && hankeYht.etunimi.isNotBlank() && hankeYht.email.isNotBlank()
-                    && hankeYht.puhelinnumero.isNotBlank()) {
-                val hankeYhtEntity = HankeYhteystietoEntity(
-                        contactType,
-                        hankeYht.sukunimi,
-                        hankeYht.etunimi,
-                        hankeYht.email,
-                        hankeYht.puhelinnumero,
-                        hankeYht.organisaatioId,
-                        hankeYht.organisaatioNimi,
-                        hankeYht.osasto,
-                        1, //TODO: real user , make sure you don't update by another user but only once
-                        hankeYht.createdAt?.toLocalDateTime(),
-                        null, //TODO: real user , make sure updated only if real changes? or what is the design decision?
-                        getCurrentTimeUTCAsLocalTime(),  //TODO: only if changed?  do we always want to change the date? how do we know has subobject really been updated?
-                        hankeYht.id,
-                        hankeEntity)
-
-                hankeEntity.listOfHankeYhteystieto!!.add(hankeYhtEntity)
+    private fun processIncomingHankeYhteystietosOfSpecificTypeToEntity(
+            listOfHankeYhteystiedot: List<HankeYhteystieto>, hankeEntity: HankeEntity, contactType: ContactType,
+            userid: Int, existingYTs: MutableMap<Int, HankeYhteystietoEntity>) {
+
+        for (hankeYht in listOfHankeYhteystiedot) {
+            // Is the incoming YT new (does not have id, create new) or old (has id, update existing):
+            if (hankeYht.id == null) {
+                // New YT, check if it is fully empty; if anything set, create new one.
+                // Note, validator should have enforced that at least the four main fields are all set (or none of them set).
+                if (hankeYht.sukunimi.isNotBlank() || hankeYht.etunimi.isNotBlank()
+                        || hankeYht.email.isNotBlank() || hankeYht.puhelinnumero.isNotBlank()
+                        || !hankeYht.organisaatioNimi.isNullOrBlank() || !hankeYht.osasto.isNullOrBlank()) {
+                    val hankeYhtEntity = HankeYhteystietoEntity(
```

Comment:
This is now allowing partially OK data to go to database. We need all 4 of them. Otherwise some other level data check has failed.
Osasto and organisaatioNimi can be blank for private citizen. I don't understand why did you add rules for them.

Reply:
This will need a bigger discussion about the logic on these (and also on other) validations; if this part works in any other way, there will too many state combinations for an easy condition expression here.